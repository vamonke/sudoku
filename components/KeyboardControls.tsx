import { useEffect } from "react";

type Props = {
  reset: () => void;
  newGame: () => void;
  undo: () => void;
  toggleShowConflict: () => void;
  showConflict: boolean;
};

export default function KeyboardControls(props: Props) {
  const { reset, newGame, undo, toggleShowConflict, showConflict } = props;

  useEffect(() => {
    const eventListener = (event: KeyboardEvent) => {
      if (event.key === "r") reset(); // TODO: Add confirmation
      if (event.key === "n") newGame(); // TODO: Add confirmation
      if (event.key === "e") toggleShowConflict();
      if (event.key === "z") undo();
    };
    window.addEventListener("keyup", eventListener);
    return () => {
      window.removeEventListener("keyup", eventListener);
    };
  }, [newGame, reset, toggleShowConflict, undo]);

  const buttons = [
    { key: "reset", text: "Reset", shortcut: "R", onClick: reset },
    { key: "new", text: "New Game", shortcut: "N", onClick: newGame },
    { key: "undo", text: "Undo", shortcut: "Z", onClick: undo },
    {
      key: "toggle",
      text: showConflict ? "Hide errors" : "Highlight errors",
      shortcut: "E",
      onClick: toggleShowConflict,
    },
  ];

  return (
    <div className="flex flex-row gap-6 items-center justify-even">
      {buttons.map((button) => (
        <button
          key={button.key}
          data-testid={button.key}
          onClick={button.onClick}
          className="flex items-center gap-2"
        >
          <div className="text-xs w-6 h-6 flex justify-center items-center bg-white shadow-sm rounded-md font-semibold border-2">
            {button.shortcut}
          </div>
          <div>{button.text}</div>
        </button>
      ))}
    </div>
  );
}
