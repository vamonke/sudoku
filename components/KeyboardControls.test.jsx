import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import KeyboardControls from "./KeyboardControls";

describe("KeyboardControls", () => {
  describe("renders correctly", () => {
    it("when showConflict is true", () => {
      const { container } = render(<KeyboardControls showConflict={true} />);
      expect(container).toMatchSnapshot();
    });

    it("when showConflict is false", () => {
      const { container } = render(<KeyboardControls showConflict={false} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe("reset function", () => {
    it("is called when the reset button is clicked", () => {
      const mockReset = jest.fn();
      render(<KeyboardControls reset={mockReset} />);
      const resetButton = screen.getByTestId("reset");
      resetButton.click();
      expect(mockReset).toHaveBeenCalled();
    });

    it("is called when r key is pressed", () => {
      const mockReset = jest.fn();
      const { container } = render(<KeyboardControls reset={mockReset} />);
      fireEvent.keyUp(container, { key: "r" });
      expect(mockReset).toHaveBeenCalled();
    });
  });

  describe("newGame", () => {
    it("is called when the new button is clicked", () => {
      const mockNewGame = jest.fn();
      render(<KeyboardControls newGame={mockNewGame} />);
      const newButton = screen.getByTestId("new");
      newButton.click();
      expect(mockNewGame).toHaveBeenCalled();
    });

    it("is called when n key is pressed", () => {
      const mockNewGame = jest.fn();
      const { container } = render(<KeyboardControls newGame={mockNewGame} />);
      fireEvent.keyUp(container, { key: "n" });
      expect(mockNewGame).toHaveBeenCalled();
    });
  });

  describe("undo function", () => {
    it("is called when the undo button is clicked", () => {
      const mockUndo = jest.fn();
      render(<KeyboardControls undo={mockUndo} />);
      const undoButton = screen.getByTestId("undo");
      undoButton.click();
      expect(mockUndo).toHaveBeenCalled();
    });

    it("is called when z key is pressed", () => {
      const mockUndo = jest.fn();
      const { container } = render(<KeyboardControls undo={mockUndo} />);
      fireEvent.keyUp(container, { key: "z" });
      expect(mockUndo).toHaveBeenCalled();
    });
  });

  describe("toggleShowConflict function", () => {
    it("is called when the toggle button is clicked", () => {
      const mockToggleShowConflict = jest.fn();
      render(<KeyboardControls toggleShowConflict={mockToggleShowConflict} />);
      const toggleButton = screen.getByTestId("toggle");
      toggleButton.click();
      expect(mockToggleShowConflict).toHaveBeenCalled();
    });

    it("is called when E key is pressed", () => {
      const mockToggleShowConflict = jest.fn();
      const { container } = render(
        <KeyboardControls toggleShowConflict={mockToggleShowConflict} />
      );
      fireEvent.keyUp(container, { key: "e" });
      expect(mockToggleShowConflict).toHaveBeenCalled();
    });
  });
});
