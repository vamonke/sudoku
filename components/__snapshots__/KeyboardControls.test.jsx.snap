// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardControls renders correctly when showConflict is false 1`] = `
<div>
  <div
    class="flex flex-row gap-4 items-center justify-even"
  >
    <button
      data-testid="reset"
    >
      Reset (R)
    </button>
    <button
      data-testid="new"
    >
      New Game (N)
    </button>
    <button
      data-testid="undo"
    >
      Undo (Z)
    </button>
    <button
      data-testid="toggle"
    >
      Highlight errors (E)
    </button>
  </div>
</div>
`;

exports[`KeyboardControls renders correctly when showConflict is true 1`] = `
<div>
  <div
    class="flex flex-row gap-4 items-center justify-even"
  >
    <button
      data-testid="reset"
    >
      Reset (R)
    </button>
    <button
      data-testid="new"
    >
      New Game (N)
    </button>
    <button
      data-testid="undo"
    >
      Undo (Z)
    </button>
    <button
      data-testid="toggle"
    >
      Hide errors (E)
    </button>
  </div>
</div>
`;
